src/xq_server.cpp

```diff
@@ -13,16 +13,16 @@
 #include <QDebug>
 #include <QTimer>
 
-xqServer::xqServer(zmq::context_t& context,
-                   const xeus::xconfiguration& c,
-                   nl::json::error_handler_t eh)
+xq_server::xq_server(zmq::context_t& context,
+                     const xeus::xconfiguration& c,
+                     nl::json::error_handler_t eh)
     :
     xserver_zmq(context, c, eh),
-    m_worker_thread(this, &m_shell, &m_controller,p_auth.get())
+    m_worker_thread(this, &m_shell, &m_controller, p_auth.get())
```

Comment:
I was reading something today about threads... And I think to be able to pass references to threads you need to use `std::ref()`? Otherwise it will just pass it as copy because threads are unaware of types.
But I might be confused? @JohanMabille or @DerThorsten? :)

Reply:
Here you are passing pointers, so the copy is fine. It would be better (regarding the ownership) to pass references if you were instantiating a more high level object AND if the memory ownership was not clear. Here, getting a pointer from a `unique_ptr` without resetting it makes clear that the thread is not responsible for the deletion of the authentication object, and assuming consistency in the code, it won't be the case for other pointers.

Anyway, no need to use `std::ref` here, your constructor could accept references and stores them as pointers:

```cpp

        WorkerThread(QObject* parent,
                     zmq::socket_t& p_shell,
                     zmq::socket_t& p_controller,
                     xeus::xauthentication& p_auth)
        : QThrad(parent)
        , p_shell(&shell)
        , p_controller(&controller)
        , p_auth(&auth)
        {
        }  
```

`std::ref` and `std::cref` are usually used in generic code where you want to be able to store references in a copyable and assignable object (meaning assigning the reference wrapper will rebind the internal ref to the new object instead of assigning the underlying object). This is equivalent to store the reference as a pointer (by taking its address as in the code above) but allow you to write generic code that handles passing references and values to functions indistinctly (whereas with pointers you would have to write a special case for dereferencing the pointer before calling the function).